Waterfall caching and redrawing

Cat-Ion opened the pull request:
I've been doing some work on #393 and it's in a mostly working state. I still don't have a way of limiting used memory right now, but at least the backend part of that should be simple enough. You do get some visual artifacts when moving the waterfall, but that can't be avoided with the 8-bit storage for FFT bins.

Memory overhead right now is 1 byte per FFT sample, over the height of the waterfall, plus 32 bytes per line.

I've messed with the plotter class for a bit, mostly by putting draw() into a few smaller functions, and also optimized its drawing function to be much faster, as it was a major bottleneck for the redrawing.

While it probably shouldn't be merged at this stage (still need the max memory stuff), I'd love some feedback and maybe tests by other people.

csete commented:
Thanks @Cat-Ion -- I will take a look at it, though it'll probably be several weeks before I get a chance. I'm in the middle of moving and my backlog is already very long.

Cat-Ion commented:
Sure, no problem. As you can see, I still have lots to do. :)

Cat-Ion commented:
An overview of the changes I made:
- created FftBuffer class that stores FFTs as arrays of 8-byte values, along with frequency and power information. It uses a ring buffer with a size set to the height of the waterfall.
- Split up the somewhat large `CPlotter::draw()` function into `drawWaterfall()` and `draw2DSpectrum()`
- `drawWaterfall()` accumulates waterfall data (`accumulateWaterfallData()`), updates the buffer, scrolls the pixmap, and calls `drawWaterfallLine()`
- `drawWaterfallLine()` uses the `CPlotter`'s FFT array for the upper line or the buffer for all older lines, gets the appropriate data, and dumps it into an RGB buffer, which is drawn onto the waterfall pixmap.
- Changing the waterfall's axes now calls the new `redrawWaterfall()`, which redraws each line.
- removed `CPlotter::m_wfbuf` - using the RGB buffer now.
- added `CPlotter::m_wfAccum` for accumulating data as floats to match the format expected by the FFT buffer.

Advantages:
- The waterfall can be redrawn when you change the center frequency or spectral power limits.
- Having a few smaller functions instead of one large drawing function is easier to handle, IMO. At least I found it rather unwieldy when I was working on it, hence the splitting up.
- Drawing into an RGB buffer first is a huge performance gain.

Current disadvantage: Redrawing full large FFTs (above 32k or so) gets a bit choppy. This should get faster when I implement partial redrawing.

alexf91 commented:
I know there are several issues addressing this pull request (e.g. #469) , but I'll just comment here.

I've been using this for a while now, and it works fine. I didn't see any crashes or other errors.
There are some minor performance issues with large FFT sizes, but these are acceptable on a regular PC.
However, some users are running gqrx on a Raspberry Pi or similar, so this might be a problem on such systems. I think making this feature configurable would be the way to go. From what I see, this is not hard to implement.

Regarding the code: Is there a reason why you are using malloc/free instead of new[]/delete[]?
